File: src/foss-extras-packagekitqt/libdetails.cpp
#include <PackageKit/Details>
#include <QMap>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <QVariant>
#include <details.h>
#include "libdetails.h"
#include "libdetails.hxx"

PackageKit__Details* PackageKit__Details_new() {
    return new PackageKit::Details();
}

PackageKit__Details* PackageKit__Details_new2(const libqt_map /* of libqt_string to QVariant* */ other) {
    QMap<QString, QVariant> other_QMap;
    libqt_string* other_karr = static_cast<libqt_string*>(other.keys);
    QVariant** other_varr = static_cast<QVariant**>(other.values);
    for (size_t i = 0; i < other.len; ++i) {
        QString other_karr_i_QString = QString::fromUtf8(other_karr[i].data, other_karr[i].len);
        other_QMap[other_karr_i_QString] = *(other_varr[i]);
    }
    return new PackageKit::Details(other_QMap);
}

libqt_string PackageKit__Details_PackageId(const PackageKit__Details* self) {
    QString _ret = self->packageId();
    // Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
    QByteArray _b = _ret.toUtf8();
    libqt_string _str;
    _str.len = _b.length();
    _str.data = static_cast<const char*>(malloc(_str.len + 1));
    memcpy((void*)_str.data, _b.data(), _str.len);
    ((char*)_str.data)[_str.len] = '\0';
    return _str;
}

libqt_string PackageKit__Details_Description(const PackageKit__Details* self) {
    QString _ret = self->description();
    // Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
    QByteArray _b = _ret.toUtf8();
    libqt_string _str;
    _str.len = _b.length();
    _str.data = static_cast<const char*>(malloc(_str.len + 1));
    memcpy((void*)_str.data, _b.data(), _str.len);
    ((char*)_str.data)[_str.len] = '\0';
    return _str;
}

int PackageKit__Details_Group(const PackageKit__Details* self) {
    return static_cast<int>(self->group());
}

libqt_string PackageKit__Details_Summary(const PackageKit__Details* self) {
    QString _ret = self->summary();
    // Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
    QByteArray _b = _ret.toUtf8();
    libqt_string _str;
    _str.len = _b.length();
    _str.data = static_cast<const char*>(malloc(_str.len + 1));
    memcpy((void*)_str.data, _b.data(), _str.len);
    ((char*)_str.data)[_str.len] = '\0';
    return _str;
}

libqt_string PackageKit__Details_Url(const PackageKit__Details* self) {
    QString _ret = self->url();
    // Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
    QByteArray _b = _ret.toUtf8();
    libqt_string _str;
    _str.len = _b.length();
    _str.data = static_cast<const char*>(malloc(_str.len + 1));
    memcpy((void*)_str.data, _b.data(), _str.len);
    ((char*)_str.data)[_str.len] = '\0';
    return _str;
}

libqt_string PackageKit__Details_License(const PackageKit__Details* self) {
    QString _ret = self->license();
    // Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
    QByteArray _b = _ret.toUtf8();
    libqt_string _str;
    _str.len = _b.length();
    _str.data = static_cast<const char*>(malloc(_str.len + 1));
    memcpy((void*)_str.data, _b.data(), _str.len);
    ((char*)_str.data)[_str.len] = '\0';
    return _str;
}

unsigned long long PackageKit__Details_Size(const PackageKit__Details* self) {
    return static_cast<unsigned long long>(self->size());
}

void PackageKit__Details_Delete(PackageKit__Details* self) {
    delete self;
}

File: src/foss-extras-packagekitqt/libdetails.zig
const QtC = @import("qt6zig");
const qtc = @import("qt6c");
const QVariant = @import("libqt6").QVariant;
const std = @import("std");
const transaction_enums = @import("libtransaction_1.zig").enums;
const ArrayMap_constu8_QVariant = std.array_hash_map.String(QVariant);

// Also inherits unprojectable QMap<QString, QVariant>

/// ### [Upstream resources](https://github.com/PackageKit/PackageKit-Qt)
pub const PackageKit__Details = extern struct {
    /// ### [Upstream resources](https://github.com/PackageKit/PackageKit-Qt)
    ///
    /// The pointer to the underlying Qt C++ object
    ///
    ptr: QtC.PackageKit__Details,

    pub const _is_PackageKit__Details = {};

    /// ### DEPRECATED: Use `new` instead
    ///
    pub const New = new;

    /// Allocate a new PackageKit::Details object in C++ memory
    ///
    pub fn new() PackageKit__Details {
        return .{ .ptr = qtc.PackageKit__Details_new() };
    }

    /// ### DEPRECATED: Use `new2` instead
    ///
    pub const New2 = new2;

    /// Allocate a new PackageKit::Details object in C++ memory
    ///
    /// ## Parameter(s):
    ///
    /// ` allocator: std.mem.Allocator `
    ///
    /// ` other: ArrayMap_constu8_QVariant `
    ///
    pub fn new2(allocator: std.mem.Allocator, other: ArrayMap_constu8_QVariant) PackageKit__Details {
        const other_count = other.count();
        const other_keys = allocator.alloc(qtc.libqt_string, other_count) catch @panic("PackageKit__Details.new2: Memory allocation failed");
        defer allocator.free(other_keys);
        const other_values = allocator.alloc(QtC.QVariant, other_count) catch @panic("PackageKit__Details.new2: Memory allocation failed");
        defer allocator.free(other_values);
        var i: usize = 0;
        var other_it = other.iterator();
        while (other_it.next()) |it_entry| : (i += 1) {
            const other_key = it_entry.key_ptr.*;
            other_keys[i] = qtc.libqt_string{
                .len = other_key.len,
                .data = other_key.ptr,
            };
            other_values[i] = @ptrCast(it_entry.value_ptr.*.ptr);
        }
        const other_map = qtc.libqt_map{
            .len = other_count,
            .keys = @ptrCast(other_keys.ptr),
            .values = @ptrCast(other_values.ptr),
        };
        return .{ .ptr = qtc.PackageKit__Details_new2(other_map) };
    }

    /// ### DEPRECATED: Use `packageId` instead
    ///
    pub const PackageId = packageId;

    /// ### [Upstream resources](https://github.com/PackageKit/PackageKit-Qt)
    ///
    /// ## Parameter(s):
    ///
    /// ` self: PackageKit__Details `
    ///
    /// ` allocator: std.mem.Allocator `
    ///
    pub fn packageId(self: PackageKit__Details, allocator: std.mem.Allocator) []const u8 {
        var _str = qtc.PackageKit__Details_PackageId(@ptrCast(self.ptr));
        defer qtc.libqt_string_free(&_str);
        const _ret = allocator.alloc(u8, _str.len) catch @panic("PackageKit__Details.packageId: Memory allocation failed");
        @memcpy(_ret, _str.data[0.._str.len]);
        return _ret;
    }

    /// ### DEPRECATED: Use `description` instead
    ///
    pub const Description = description;

    /// ### [Upstream resources](https://github.com/PackageKit/PackageKit-Qt)
    ///
    /// ## Parameter(s):
    ///
    /// ` self: PackageKit__Details `
    ///
    /// ` allocator: std.mem.Allocator `
    ///
    pub fn description(self: PackageKit__Details, allocator: std.mem.Allocator) []const u8 {
        var _str = qtc.PackageKit__Details_Description(@ptrCast(self.ptr));
        defer qtc.libqt_string_free(&_str);
        const _ret = allocator.alloc(u8, _str.len) catch @panic("PackageKit__Details.description: Memory allocation failed");
        @memcpy(_ret, _str.data[0.._str.len]);
        return _ret;
    }

    /// ### DEPRECATED: Use `group` instead
    ///
    pub const Group = group;

    /// ### [Upstream resources](https://github.com/PackageKit/PackageKit-Qt)
    ///
    /// ## Parameter(s):
    ///
    /// ` self: PackageKit__Details `
    ///
    /// ## Returns:
    ///
    /// ` transaction_enums.Group `
    ///
    pub fn group(self: PackageKit__Details) i32 {
        return qtc.PackageKit__Details_Group(@ptrCast(self.ptr));
    }

    /// ### DEPRECATED: Use `summary` instead
    ///
    pub const Summary = summary;

    /// ### [Upstream resources](https://github.com/PackageKit/PackageKit-Qt)
    ///
    /// ## Parameter(s):
    ///
    /// ` self: PackageKit__Details `
    ///
    /// ` allocator: std.mem.Allocator `
    ///
    pub fn summary(self: PackageKit__Details, allocator: std.mem.Allocator) []const u8 {
        var _str = qtc.PackageKit__Details_Summary(@ptrCast(self.ptr));
        defer qtc.libqt_string_free(&_str);
        const _ret = allocator.alloc(u8, _str.len) catch @panic("PackageKit__Details.summary: Memory allocation failed");
        @memcpy(_ret, _str.data[0.._str.len]);
        return _ret;
    }

    /// ### DEPRECATED: Use `url` instead
    ///
    pub const Url = url;

    /// ### [Upstream resources](https://github.com/PackageKit/PackageKit-Qt)
    ///
    /// ## Parameter(s):
    ///
    /// ` self: PackageKit__Details `
    ///
    /// ` allocator: std.mem.Allocator `
    ///
    pub fn url(self: PackageKit__Details, allocator: std.mem.Allocator) []const u8 {
        var _str = qtc.PackageKit__Details_Url(@ptrCast(self.ptr));
        defer qtc.libqt_string_free(&_str);
        const _ret = allocator.alloc(u8, _str.len) catch @panic("PackageKit__Details.url: Memory allocation failed");
        @memcpy(_ret, _str.data[0.._str.len]);
        return _ret;
    }

    /// ### DEPRECATED: Use `license` instead
    ///
    pub const License = license;

    /// ### [Upstream resources](https://github.com/PackageKit/PackageKit-Qt)
    ///
    /// ## Parameter(s):
    ///
    /// ` self: PackageKit__Details `
    ///
    /// ` allocator: std.mem.Allocator `
    ///
    pub fn license(self: PackageKit__Details, allocator: std.mem.Allocator) []const u8 {
        var _str = qtc.PackageKit__Details_License(@ptrCast(self.ptr));
        defer qtc.libqt_string_free(&_str);
        const _ret = allocator.alloc(u8, _str.len) catch @panic("PackageKit__Details.license: Memory allocation failed");
        @memcpy(_ret, _str.data[0.._str.len]);
        return _ret;
    }

    /// ### DEPRECATED: Use `size` instead
    ///
    pub const Size = size;

    /// ### [Upstream resources](https://github.com/PackageKit/PackageKit-Qt)
    ///
    /// ## Parameter(s):
    ///
    /// ` self: PackageKit__Details `
    ///
    pub fn size(self: PackageKit__Details) usize {
        return qtc.PackageKit__Details_Size(@ptrCast(self.ptr));
    }

    /// ### DEPRECATED: Use `delete` instead
    ///
    pub const Delete = delete;

    /// ### [Upstream resources](https://github.com/PackageKit/PackageKit-Qt)
    ///
    /// Delete this object from C++ memory
    ///
    /// ## Parameter:
    ///
    /// ` self: PackageKit__Details `
    ///
    pub fn delete(self: PackageKit__Details) void {
        qtc.PackageKit__Details_Delete(@ptrCast(self.ptr));
    }
};
